### PR TITLE
[iOS] Animated image improvements

### PR DESCRIPTION
**Pull request by ericlewis**
## Summary
The goal of this PR is to improve the pipeline currently used for displaying GIFs / animated images on iOS. It is achieved by not holding all of the decoded frames in memory at the same time, as well as happily releasing existing memory whenever possible. This code is a simplified version of what you would find in SDWebImage (it is nearly 1:1, with unsupported or uneeded things removed). By adopting this API, it also allows classes conforming to RCTImageURLLoader or RCTImageDataDecoder to return any decodable UIImages conforming to RCTAnimatedImage and have improvements to memory consumption. Because RCTAnimatedImage is just a subset of the SDAnimatedImage protocol, it also means that you can use SDWebImage easier with Image directly.

A nice to have would be progressive image loading, but is beyond scope for this PR. It would, however, touch most of these same parts.

## Changelog

[iOS] [Fixed] - Substantially lower chances of crashes from abundant GIF use

## Test Plan
TBD. (but i am running a version of this in my own app currently)

### CODE REVIEW

**Comment by ericlewis** [comment]
This doesn't necessarily fix overall memory usage currently (lowers it, but not much), but prevents crashes from OOM when using lots of GIFs & also conforms to a similar API as SDAnimatedImage. Once I fix the initial flash and look in to the overall memory usage a bit- will flip to acceptance!

**Comment by ericlewis** [comment]
Coming up with test cases now, the memory pressure issue seems resolved. I may post a video demonstrating how an app crashes with lots of gif's on master, but how it can continue to run using this PR. Overall memory consumption is the same as master for light loads.

**Comment by ericlewis** [comment]
@zhongwuzw would you mind reviewing this?

**Comment by zhongwuzw** [comment]
@ericlewis Good job, I'll find some time to review this.

File: Libraries/Image/RCTAnimatedImage.m [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  }
+}
+
+- (void)dealloc
+{
+  if (_imageSource) {
+    CFRelease(_imageSource);
+    _imageSource = NULL;
+  }
+  [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidReceiveMemoryWarningNotification object:nil];
```
Comment (sammy-SC) [inline Libraries/Image/RCTAnimatedImage.m]: Is this necessary on iOS 9+?

**Comment by sammy-SC** [comment]
@ericlewis nice work, by any chance have you come up with the test cases?

**Comment by ericlewis** [comment]
Easiest test case is dumping a TON of gifs on a single page and noticing it’s not crashing. It’s harder to do this, however, purely in simulator. And the results can only truly be realized on devices. 

Also, the way observers are used is integral to how this functions. 

Sent via Superhuman iOS ( https://sprh.mn/?vip=ericlewis777@gmail.com )

**Comment by shergin** [comment]
cc @janicduplessis

**Comment by shergin** [comment]
Eric, should we land this? Is this polish enough? ;)

File: Libraries/Image/RCTAnimatedImage.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#import <UIKit/UIKit.h>
+
+@protocol RCTAnimatedImage <NSObject>
+@property (nonatomic, assign, readonly) NSUInteger animatedImageFrameCount;
+@property (nonatomic, assign, readonly) NSUInteger animatedImageLoopCount;
+
+- (nullable UIImage *)animatedImageFrameAtIndex:(NSUInteger)index;
```
[the remark this answers is not in the document]
Comment (ericlewis) [reply Libraries/Image/RCTAnimatedImage.h]: Indeed. I should add it.

**Comment by ericlewis** [comment]
Technically they do allow it, but I also feel there is sufficient changes to how it actually works too. I think it is pretty good, at the end of the day it’s doing more good than bad :)

**Comment by sahrens** [comment]
Looks like this might have introduced a memory leak. We're still investigating to verify.

**Pull request closed without merging**